### PR TITLE
Adding jury-rigged frames to the game

### PR DESCRIPTION
**Pull request by LilimDiam**
#### Summary
Balance "Make players unable to install proper frames, instead letting them install less durable variants"

#### Purpose of change
Making it harder for players to create death-mobiles, and forcing them to rely on preexistent frame structures rather than trying to enhance them.

#### Describe the solution
Made players unable to install extra light, heavy duty and normal frames, replacing them with variants that have about 2/3 durability and damage resistance as originals.

#### Describe alternatives you've considered
Making it possible to install proper frames using something like a car assembly line.

#### Testing
Can repair old frames, can remove them. Installing new frames is only possible via jury-rigging.

#### Additional context
N/A

### CODE REVIEW

**Comment by AtomicFox556** [comment]
I don't think it makes much sense for extra-light frames at least; such frames themselves consist of pipes welded together, and so there wouldn't be a perceptible difference in durability if they were built on factory equipment or welded together by someone who is competent enough at welding.

Furthermore, seems like forge welding could be used to join pieces of metal together with the joint being very close in strength to the material itself, if someone wanted to bother doing this to make a vehicle frame.

**Comment by LilimDiam** [comment]
I was on the fence about extra-light frames myself, although I would need to look up in real life analogs to be able to tell for sure.
As for forge welding, fair, but I don't see survivors having access to a forge suitable for making a car frame from scratch or enhancing existing one.

**Comment by LilimDiam** [comment]
Ok so from what I was able to find, aluminium car frames do exist, but their production method isn't that dissimilar to steel frames. There is an obvious discrepancy with the way you can make extra light frames in-game from pipes, but I don't think that warrants a change to this PR, rather it should require a change to recipe, if anything

**Comment by PatrikLundell** [comment]
This plays very poorly with the recently introduced nerf that causes everything to degrade when damaged, so even if you somehow managed to find a somewhat suitable vehicle in perfect condition you'd eventually have to replace the frames of in with your new crap ones as they can no longer be repaired to a level that's better than crap frames (unless you somehow managed to never run into anything at all, and kept zombies away from damaging it as well).

In addition to this, it also means that all existing frames are now just trash to be recycled into crap frames, which would take a lot of extra time to first dismantle the frames and then construct replacement crap frames from scratch.

It can be noted that in-game aluminum frames crumble if you sneeze at them, so even weaker crap variants would be useful only if you're prepared to walk instead if the vehicle happened to run into a chicken.

**Comment by LilimDiam** [comment]
You would still be able to repair an XX frame by replacing it with a brand new one, although I'm not sure if this should be possible. 

Also not sure what you mean by saying all existing frames are trash to be recycled, they can still be used for everything other than enhancing pre-existing frames. Which, yes, is intentional.

**Comment by Maleclypse** [comment]
So if I remember correctly from some of the conversations around this topic, the jury rigged heavy frames might not need to have a small hp and damage resistance but instead would need to have a multiple of the old world frames weight.   That basically the survivor can build a tough battle vehicle but it's gonna have awful fuel efficiency.

**Comment by AtomicFox556** [comment]
> As for forge welding, fair, but I don't see survivors having access to a forge suitable for making a car frame from scratch or enhancing existing one.

Obviously a coal forge isn't really suitable for this purpose as you can't stuff an object that large into a coal pile, but an oxyacetylene torch should work if you don't mind very high fuel consumption.

**Comment by PatrikLundell** [comment]
As far as I know, you can only "repair" a vehicle part by replacement if the part is in the "XX" state, so you'd have to either drive along with a vehicle with lots of parts on the verge of destruction and hope none of them suffer more than one damaging blows, or replace them with crap frames in a timely manner.

It can also be noted that it also would break the current logic that has "repair" of a destroyed vehicle part being a shorthand equivalent of removing it and inserting a new part of the same type.

**Comment by AtomicFox556** [comment]
> Fair point on the oxy torch, how much fuel consumption are we talking about here? Something in the vein of 4-5 times the amount it takes to just weld something in place?

This seems close enough. It's supposed to be a very inefficient (but possible) method of getting things done.

**Comment by I-am-Erk** [comment]
I don't think we're fully opposed to better homemade frames being available, locked behind sufficient tools and proficiencies, but those are a separate issue. I highly doubt your friend is making replacement parts from scratch out of pipes and scrap metal: he's getting them from a supplier. He's also an extremely skilled person with a well supplied shop.

Generally the process for a change like this is:
1. Move to the initial phase, stuff a survivor with rudimentary skills will be able to do.
2. Unlock better parts and processes for higher tier survivors to access

We're looking at 1 right now.

If you've got some good references to suggest that, for example, we should maintain the generic commercial "frame" part as something that can be installed as an agnostic replacement on all vehicles, I think we'd be open to that, and just make it non-craftable unless you have sufficient skill to do so. however I'm under the impression that if you want to replace the front end of a 2019 Coyota Tamry, you'll need a front end from another similar model Coyota Tamry, or at least something very close and a ton of skill and equipment. For that, we'd need to hugely expand the item list of what frames are considered appropriate. Just using a frame you pulled off a shopping buggy will not suffice... and currently it would.

**Comment by I-am-Erk** [comment]
To expand on what would be needed to make sense, if we want to have commercial frame replacements then we'll need to do at least things like having "sedan frame front end" that can be harvested and put on a sedan, and if you try to use a sedan frame front end on a pickup or a semi, it is inadequate. This is well outside the scope of this PR and I don't think it should be a blocker for this PR.

**Comment by keampe** [comment]
Hmmm... There's a lot there - it depends on the depth you want to go. For instance most games would have the light frames be easier than the heavy frames but in cdda the extra light frames are aluminum. Aluminum is harder to weld or at least takes a slightly different skill set. Also, there's proficiency's to take into consideration. IRL things like welding are most often used on buildings and vehicles but currently there is no usage of proficiency's like welding in either the construction menu or the vehicle menu. This might be an opportune time to introduce them.

So, without proficiency (might be out of scope for this PR) I'd say raise the Mechanics requirement by at least 2 and add in a Fabrication of 4 or 6. Basic welding gets you to 4 and advanced welding get you to 6, so I'm inclined to think 4. It's basic welding - not really fancy and doesn't have to look pretty. OTOH, you're most often not just installing the frame but adapting it (almost always coming from a different vehicle), so maybe 6 is more appropriate.

Out of scope for this PR, but something that should probably be looked into is the adding of frames to an existing chassis or creating one from scratch. Some sort of design proficiency should be involved there.

   - Shane

**Comment by I-am-Erk** [comment]
All of this is out of the scope of this PR. This PR accomplishes a primary goal, which is to make it not trivially easy to just make your car into a 30' long, 20' wide mobile superfortress by welding on scrap metal you harvested from random wrecks or dug up in a garbage dump.

Figuring out an appropriate system to repair and refit commercial automotive chassis back to their original spec if you find the right parts or have the skill to make the wrong parts work is a great potential addition to the game, but it's a totally different thing from this PR.

File: data/json/vehicleparts/frames.json [hunk 1]
```diff
@@ -56,10 +56,29 @@
...
     "damage_reduction": { "all": 20 }
   },
+  {
+    "id": "jrframe",
+    "type": "vehicle_part",
+    "name": { "str": "jury-rigged frame" },
+    "item": "frame",
+    "location": "structure",
+    "standard_symbols": true,
+    "durability": 250,
```
Comment (I-am-Erk) [inline data/json/vehicleparts/frames.json:70]: Jury rigged frames should actually be a bit tougher than commercial frames, at the significant cost of weight. Later, we'll also need to make them transfer damage to other parts of the car and the occupants, so if you're using a jury rigged vehicle frame you lose the benefit of a crumple zone.

**Comment by PatrikLundell** [comment]
I'd be careful about making crap frames heavier, because the game has removed the ability to add more than two engines to a vehicle, so you have a very real limit there.

My 4*6 "deathmobile" built out of the sturdiest parts I can find (i.e. heavy frames, etc.), has one enhanced electric engine and one V8 diesel engine, and is capable of towing security vans, humvees, etc. with both engines engaged, provided the vehicle isn't fully loaded, but fails on plains when fully loaded with heavy stuff in 1½ cubic meters of storage.
This isn't a large vehicle, and there isn't that much more room for larger engines (2 12 cylinder diesel engines at tank vortex inducing fuel consumption in the realistic maximum, I think (there are damaged jet engines that may or may not be possible to be salvaged)).

**Comment by I-am-Erk** [comment]
> I'd be careful about making crap frames heavier, because the game has removed the ability to add more than two engines to a vehicle, so you have a very real limit there.
> 
> My 4*6 "deathmobile" built out of the sturdiest parts I can find (i.e. heavy frames, etc.), has one enhanced electric engine and one V8 diesel engine, and is capable of towing security vans, humvees, etc. with both engines engaged, provided the vehicle isn't fully loaded, but fails on plains when fully loaded with heavy stuff in 1½ cubic meters of storage. This isn't a large vehicle, and there isn't that much more room for larger engines (2 12 cylinder diesel engines at tank vortex inducing fuel consumption in the realistic maximum, I think (there are damaged jet engines that may or may not be possible to be salvaged)).

That's a relevant point to consider, ie. that homemade frames shouldn't be so heavy that you can't make a vehicle function at all, but requiring players to carefully consider weight tradeoffs and worry about risking damage to their lighter-weight vehicle frames if they plow through too many zombies is absolutely an intentional part of the design here.

File: data/json/vehicleparts/frames.json (hunk 1 above)
Comment (LilimDiam) [reply data/json/vehicleparts/frames.json:70]: The issue with that is that just doing so would violate mass preservation, as somehow a singular frame weights more when jury-rigged
That can be worked around by making dedicated recipes for jury-rigged frames, but that's not quite what I had in mind when making this PR

**Comment by kevingranade** [comment]
Some variant of this is what we are doing, the only question is when and details.

If you look at the whole process we're representing here, the status quo is repeatedly crashing a vehicle and repairing it back to "like new" condition, which is not reasonable.
We don't model the individual processes that make that the case such as metal fatigue, frame warping, micro cracks, but we are attempting to represent that outcome.

> I don't think it makes much sense for extra-light frames at least; such frames themselves consist of pipes welded together, 

Good point, the target here is car frames, not bicycle or go kart frames.

> forge welding

AFAIK this is not applicable for vehicle scale structural beams.  The size is a problem, the materials are a problem, and the fact that it's a finished product is a problem.

> This plays very poorly with the recently introduced nerf that causes everything to degrade when damaged, so even if you somehow managed to find a somewhat suitable vehicle in perfect condition you'd eventually have to replace the frames of in with your new crap ones as they can no longer be repaired to a level that's better than crap frames (unless you somehow managed to never run into anything at all, and kept zombies away from damaging it as well).

This is working as intended, vehicles are precious.  There are certainly some issues remaining around this such as vanilla zombies being able to damage vehicle frames in the 1st place, But that kind of thing is exactly the point of a change like this, it exposes the other problems in the game that are tolerable and therefore ignored.

> You would still be able to repair an XX frame by replacing it with a brand new one, although I'm not sure if this should be possible.

It should not.

> Find the factory weld and grind off the broken frame there.
> Weld the new frame where the factory would have.

Yes you can do this, but how many times? Ideally the frame repair process would consume matching frames, and the result would feed into our durability erosion system.

> Take a look at any *classic car* rally.

This is survivorship bias. You see cars that had repairable damage. You don't see the cars that were damaged too far to be worth salvaging and if you repeatedly wreck and repair a car they will always degrade to that point.

> Most people these days don't have that reason - but your survivor does.

That's where you're wrong. The survivor has no compelling incentive to put this much work into repairing *the same vehicle* over and over again. As you outline, if you're reparing frames you have to disassemble the car partially in order to do it, at which point migrating your customizations and components to another car becomes the easier option at some point and if your survivor is being pragmatic that point is going to approach very rapidly. Thebproblem here is you're applying vanity project effort to a survival situation and that does not line up.

> I'd be careful about making crap frames heavier, because the game has removed the ability to add more than two engines to a vehicle, so you have a very real limit there.

Again, this is fully intended. Want a huge vehicle? Strip down a bus or other large vehicle and build it up.

File: data/json/vehicleparts/frames.json (hunk 1 above)
Comment (LilimDiam) [reply data/json/vehicleparts/frames.json:70]: To elaborate, I think having a dedicated recipe for combining, say, 2 heavy duty frames into jury-rigged one *would* make sense, but I don't want to add this to all kinds of frames as this just adds more keypresses for player without really good reason, at least imo
Is there a way for vehicle part to require multiple items, btw? If yes, that would be quite simple solution to that

**Comment by LilimDiam** [comment]
> Compression zones etc. are great for protecting lives in a collision, but when the creatures you collide with are undead you don't want to protect them from harm, and you definitely don't want to damage your vehicle doing it.

I'm hardly an expert, but aren't crumple zones designed to protect, primarily, people that are inside the car during crash? I'm not sure I would like to be in a car with rigid frame going for a head-on collision with a hulk or something along the lines.

**Comment by kevingranade** [comment]
> Well, "too many" is approximately one gotcha graboid. Any collision with those buggers at marching speed will require repair and probably permanently damaged parts, with shock absorbers being fairly good at protecting the interior.

Yep, working as intended, except the shock absorbers, which are kind of nonsense.

> I think it's possible to do limited zombie plowing at lower speeds with limited damage, but haven't dared to try it after the permanent damage introduction. I wouldn't use regular frames for anything that's supposed to survive a collision with anything (they can sometimes hit shrubs without taking damage).

The answer to this is to establish the parameters of a collision that should not cause car damage, or more generally characterize what kind of damage we should expect from a particular crash scenario, and test for that, not reject only vaguely related changes because of how unintended interactions in one part of the game interact with a new feature. Collision damage numbers and outcomes are in fact all over the place, the solution isn't to keep everything super forgiving to paper over that, it's to fix the collision numbers and outcomes.

> I disagree with @kevingranade on repairing a custom vehicle. The whole reason I want a custom vehicle is because pre cataclysm vehicles aren't made for the post cataclysm world. Compression zones etc. are great for protecting lives in a collision, but when the creatures you collide with are undead you don't want to protect them from harm, and you definitely don't want to damage your vehicle doing it.

As per above, what you want is a ram bar or similar that is either anchored to the frame either rigidly or via an elastic linkage, and a test that outlines some scenarios where vehicles with ram bars collide with various targets at various speeds. In the other direction it would be nice to have explicit components representing "crumple zone" mechanics that do absorb shock energy at the cost of themselves being fragile. This is most definitely something you can add to an existing vehicle, but at the tradeoff of reducing your vehicle's survivability when impacting something much more massive than a zombie.

These kinds of parts are orthogonal to the frame issue.

> If repairing the custom vehicle becomes so expensive that you'd have to make a new one, then I'd do that rather than use a damaged standard vehicle that falls apart at the next collision with something hostile, leaving you stranded and surrounded. If a security van was up to the task I wouldn't create a custom vehicle (although I'd still lament that it's too bulky and unwieldy), and even if it was, it will fall apart due to damage since you can't repair it any longer. Good luck finding two suitable vehicles in pristine, or even good condition.

Some of these scenarios absolutely should work, and they don't because everything gets papered over with a tedious repair cycle, so no one has any incentive to fix them.

Ordinary street cars should get wrecked by plowing through zombie hordes. Heavier trucks should handle that just fine, but still be subject to lasting damage when impacting other vehicles, walls, or massive creatures. Truly overbuilt vehicles like APCs should should be able to plow through practically everything, but have *meaningful* drawbacks around how massive they are and the difficulty/impossibility of repairing them.

**Comment by PatrikLundell** [comment]
I don't have a problem with street cars getting wrecked, as that's what I would expect (and also the reason I only use those early on and then only for transport).
Heavy vehicles are still on the flimsy side. They can take a zombie or two, but probably take damage in the process.
The military vehicles are currently a lot flimsier than the target described.

File: data/json/vehicleparts/frames.json (hunk 1 above)
Comment (I-am-Erk) [reply data/json/vehicleparts/frames.json:70]: That could be done as a follow up, then; I think the jury-rigged frame *should* be a separate item with higher weight.

**Pull request closed without merging**